datafusion-cli/src/main.rs

```diff
@@ -252,26 +253,128 @@ fn is_valid_memory_pool_size(size: &str) -> Result<(), String> {
+            return Err(format!(
+                "Negative memory pool size value is not allowed '{}'",
+                size
+            ));
         }
+
+        let num = num_str.parse::<usize>().map_err(|_| {
+            format!("Invalid numeric value in memory pool size '{}'", size)
+        })?;
+
+        let suffix = caps.get(2).map(|m| m.as_str()).unwrap_or("b");
+        let unit = byte_suffixes()
+            .get(suffix)
+            .ok_or_else(|| format!("Invalid memory pool size '{}'", size))?;
+
+        Ok(num * unit.multiplier())
     } else {
-        return Err(format!("Invalid memory pool size '{}'", size));
-    };
+        Err(format!("Invalid memory pool size '{}'", size))
+    }
+}
 
-    match size.parse::<usize>() {
-        Ok(size) if size > 0 => Ok(factor * size),
-        _ => Err(format!("Invalid memory pool size '{}'", size)),
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn assert_conversion(input: &str, expected: Result<usize, String>) {
+        let result = extract_memory_pool_size(input);
+        match expected {
+            Ok(v) => assert_eq!(result.unwrap(), v),
+            Err(e) => assert_eq!(result.unwrap_err(), e),
+        }
+    }
+
+    #[test]
+    fn memory_pool_size() -> Result<(), String> {
+        // Test basic sizes without suffix, assumed to be bytes
+        assert_conversion("5", Ok(5));
+        assert_conversion("100", Ok(100));
+
+        // Test various units
+        assert_conversion("5b", Ok(5));
+        assert_conversion("4k", Ok(4 * 1024));
+        assert_conversion("4kb", Ok(4 * 1024));
+        assert_conversion("20m", Ok(20 * 1024 * 1024));
+        assert_conversion("20mb", Ok(20 * 1024 * 1024));
+        assert_conversion("2g", Ok(2 * 1024 * 1024 * 1024));
+        assert_conversion("2gb", Ok(2 * 1024 * 1024 * 1024));
+        assert_conversion("3t", Ok(3 * 1024 * 1024 * 1024 * 1024));
+        assert_conversion("4tb", Ok(4 * 1024 * 1024 * 1024 * 1024));
+
+        // Test case insensitivity
+        assert_conversion("4K", Ok(4 * 1024));
+        assert_conversion("4KB", Ok(4 * 1024));
+        assert_conversion("20M", Ok(20 * 1024 * 1024));
+        assert_conversion("20MB", Ok(20 * 1024 * 1024));
+        assert_conversion("2G", Ok(2 * 1024 * 1024 * 1024));
+        assert_conversion("2GB", Ok(2 * 1024 * 1024 * 1024));
+        assert_conversion("2T", Ok(2 * 1024 * 1024 * 1024 * 1024));
+
+        // Test invalid input
+        assert_conversion(
+            "invalid",
+            Err("Invalid memory pool size 'invalid'".to_string()),
+        );
+        assert_conversion("4kbx", Err("Invalid memory pool size '4kbx'".to_string()));
+        assert_conversion(
+            "-20mb",
+            Err("Negative memory pool size value is not allowed '-20mb'".to_string()),
+        );
+        assert_conversion(
+            "-100",
+            Err("Negative memory pool size value is not allowed '-100'".to_string()),
+        );
+
```

Comment:
```suggestion
        assert_conversion(
            "12k12k",
            Err("Invalid memory pool size '12k12k'".to_string()),
        );
```

Reply:
This case should have failed, but it passed.

Reply:
I'm not sure if I understand you correctly. Do you mean the case "-100" shouldn't fail with the Err `"Negative memory pool size value is not allowed '-100'"`?

Reply:
No, we should add a test for cases like `12k12k `.

Reply:
Nice catch! Fixed in the latest commit.